lightning/src/util/ser_macros.rs

```diff
@@ -9,35 +9,23 @@
 
 macro_rules! encode_tlv {
 	($stream: expr, {$(($type: expr, $field: expr)),*}) => { {
-		use util::ser::{BigSize, LengthCalculatingWriter};
+		use util::ser::BigSize;
 		$(
 			BigSize($type).write($stream)?;
-			let mut len_calc = LengthCalculatingWriter(0);
-			$field.write(&mut len_calc)?;
-			BigSize(len_calc.0 as u64).write($stream)?;
-			$field.write($stream)?;
+			$stream.write_all_with_length(&$field, 0)?;
 		)*
 	} }
 }
 
 macro_rules! encode_varint_length_prefixed_tlv {
 	($stream: expr, {$(($type: expr, $field: expr)),*}) => { {
-		use util::ser::{BigSize, LengthCalculatingWriter};
-		let mut len = LengthCalculatingWriter(0);
-		{
-			$(
-				BigSize($type).write(&mut len)?;
-				let mut field_len = LengthCalculatingWriter(0);
-				$field.write(&mut field_len)?;
-				BigSize(field_len.0 as u64).write(&mut len)?;
-				len.0 += field_len.0;
-			)*
-		}
-
-		BigSize(len.0 as u64).write($stream)?;
-		encode_tlv!($stream, {
+		use util::ser::{BigSize, Writer};
+		let mut buf = Writer::new();
+		encode_tlv!(&mut buf, {
 			$(($type, $field)),*
 		});
+		BigSize(buf.len() as u64).write($stream)?;
```

Comment:
Can we not do the same optimization as `encode_tlv` here? At least it would drop one more `memcpy`.

Reply:
Ah, good point.  Done with slight improvement to the crate-local interface of `Writer`.